File: frontend/src/utils/jwt.ts
export function jwtSubject(token: string): string | null {
  const parts = token.split('.')
  if (parts.length !== 3) return null

  const payload = parts[1]
  const normalized = payload.replace(/-/g, '+').replace(/_/g, '/')
  const padded = normalized + '='.repeat((4 - (normalized.length % 4)) % 4)

  try {
    const json = JSON.parse(atob(padded))
    const sub = json?.sub
    return typeof sub === 'string' ? sub : null
  } catch {
    return null
  }
}

File: lib/core/jwt.dart
import 'dart:convert';

String? jwtSubject(String token) {
  final parts = token.split('.');
  if (parts.length != 3) return null;
  final payload = parts[1];

  String normalized = payload.replaceAll('-', '+').replaceAll('_', '/');
  switch (normalized.length % 4) {
    case 0:
      break;
    case 2:
      normalized += '==';
      break;
    case 3:
      normalized += '=';
      break;
    default:
      return null;
  }

  try {
    final decoded = utf8.decode(base64.decode(normalized));
    final map = jsonDecode(decoded);
    if (map is Map<String, dynamic>) {
      final sub = map['sub'];
      return sub is String ? sub : null;
    }
  } catch (_) {
    return null;
  }
  return null;
}
